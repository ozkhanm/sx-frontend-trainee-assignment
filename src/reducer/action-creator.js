const ActionType = {
    GET_ARTICLES: `GET_ARTICLES`,
    CHANGE_LOADING_STATUS: `CHANGE_LOADING_STATUS`
};

const ActionCreator = {
    getArticles: (articles) => ({
        type: ActionType.GET_ARTICLES,
        payload: articles
    }),
    changeLoadingStatus: (status) => ({
        type: ActionType.CHANGE_LOADING_STATUS,
        payload: status
    })
};

export {ActionCreator, ActionType};