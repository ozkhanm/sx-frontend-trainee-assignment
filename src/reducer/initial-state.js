const initialState = {
    articles: [],
    isDataLoaded: false
};

export default initialState;