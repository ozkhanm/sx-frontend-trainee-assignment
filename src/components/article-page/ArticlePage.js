import React from "react";
import {connect} from "react-redux";

import PageHeader from "../page-header/PageHeader";
import LoadingPage from "../loading-page/LoadingPage";
import Comments from "../comments/Comments";
import ErrorPage from "../error-page/ErrorPage";

import {getDate} from "../../utils";
import {ActionCreator} from "../../reducer/action-creator";

const ArticlePage = (props) => {
    const {activeArticle, isActiveArticleLoaded, changeActiveArticleId} = props;

    if (isActiveArticleLoaded) {
        changeActiveArticleId(activeArticle.id);

        return (
            <React.Fragment>
                <PageHeader page={`ARTICLE_PAGE`}/>
                <div className="item-container">
                    <h1>{activeArticle.title} {activeArticle.url ? <a href={activeArticle.url} className="article-data">({activeArticle.url})</a> : ``}</h1>
                    <div className="article-data-block">
                        <p className="article-data">by: {activeArticle.by},</p>
                        <p className="article-data">posted: {getDate(activeArticle.time)},</p>
                        <p className="article-data">comments: {activeArticle.descendants}</p>
                    </div>
                    <hr/>
                    <Comments/>
                </div>
            </React.Fragment>
        );
    } else if (activeArticle === -1) {
        return <ErrorPage/>;
    } else {
        return <LoadingPage/>;
    }
};

const mapStateToProps = (state) => ({
    activeArticle: state.activeArticle,
    isActiveArticleLoaded: state.isActiveArticleLoaded
});

const mapDispatchToProps = (dispatch) => ({
    changeActiveArticleId(articleId) {
        dispatch(ActionCreator.changeActiveArticleId(articleId));
    }
});

export {ArticlePage};
export default connect(mapStateToProps, mapDispatchToProps)(ArticlePage);