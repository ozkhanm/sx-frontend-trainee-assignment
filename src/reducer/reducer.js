import initialState from "./initial-state";
import {ActionType, ActionCreator} from "./action-creator";

const Operation = {
    getArticles: () => (dispatch, getState, api) => {
        return api.get(`/newstories.json?print=pretty`)
            .then((response) => {
                const ids = response.data.slice(0, 100);
                
                return Promise.all(ids.map((it) => api.get(`item/${it}.json?print=pretty`)));
            })
            .then((response) => {
                const articles = response.slice().map((it) => it.data);

                dispatch(ActionCreator.getArticles(articles));
            });
    }
};

const reducer = (state = initialState, action) => {
    switch (action.type) {
        case ActionType.CHANGE_LOADING_STATUS:
            return Object.assign({}, state, {
                isDataLoaded: action.payload
            });

        case ActionType.GET_ARTICLES:
            return Object.assign({}, state, {
                articles: action.payload
            });
    }

    return state;
};

export {reducer, Operation};