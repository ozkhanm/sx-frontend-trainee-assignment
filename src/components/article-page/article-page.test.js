import React from "react";
import renderer from "react-test-renderer";
import {Router} from "react-router-dom";
import {Provider} from "react-redux";
import {createStore} from "redux";

import {ArticlePage} from "./ArticlePage";

import history from "../../history";
import {reducer} from "../../reducer/reducer";

it(`Article screen renders correctly if data is fetching`, () => {
    const tree = renderer
        .create(
            <Provider store={createStore(reducer)}>
                <ArticlePage isActiveArticleLoaded={false}/>
            </Provider>)
        .toJSON();
  
    expect(tree).toMatchSnapshot();
});

it(`Article screen renders correctly if data is loaded`, () => {
    const article = {
        "by" : "idletom",
        "descendants" : 0,
        "id" : 30406473,
        "score" : 1,
        "time" : 1645371191,
        "title" : "Technical Analysis of the DDoS Attacks Against Ukrainian Websites",
        "type" : "story",
        "url" : "https://www.cadosecurity.com/technical-analysis-of-the-ddos-attacks-against-ukrainian-websites/"
    };
    const tree = renderer
        .create(
            <Provider store={createStore(reducer)}>
                <Router history={history}>
                    <ArticlePage isActiveArticleLoaded={true} activeArticle={article} changeActiveArticleId={() => {}}/>
                </Router>
            </Provider>)
        .toJSON();
  
    expect(tree).toMatchSnapshot();
});

it(`Article screen renders correctly if error occured`, () => {
    const tree = renderer
        .create(
            <Provider store={createStore(reducer)}>
                <Router history={history}>
                    <ArticlePage isActiveArticleLoaded={false} activeArticle={-1} changeActiveArticleId={() => {}}/>
                </Router>
            </Provider>)
        .toJSON();
  
    expect(tree).toMatchSnapshot();
});