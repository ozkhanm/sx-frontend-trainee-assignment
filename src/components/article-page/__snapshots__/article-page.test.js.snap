// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article screen renders correctly if data is fetching 1`] = `
Array [
  <div>
    <div
      className="header-controls"
    >
      <a
        className="header-controls-button header-controls-refresh-button "
        onClick={[Function]}
      />
    </div>
    <p
      className="header"
    >
      Hacker News
    </p>
  </div>,
  <h1
    className="load"
  >
    Загрузка данных...
  </h1>,
]
`;

exports[`Article screen renders correctly if data is loaded 1`] = `
Array [
  <div>
    <div
      className="header-controls"
    >
      <a
        className="header-controls-button header-controls-refresh-button "
        onClick={[Function]}
      />
      <a
        className="header-controls-button header-controls-return-button"
        href="/"
        onClick={[Function]}
      />
    </div>
    <p
      className="header"
    >
      Hacker News
    </p>
  </div>,
  <div
    className="item-container"
  >
    <h1>
      Technical Analysis of the DDoS Attacks Against Ukrainian Websites
       
      <a
        className="article-data"
        href="https://www.cadosecurity.com/technical-analysis-of-the-ddos-attacks-against-ukrainian-websites/"
      >
        (
        https://www.cadosecurity.com/technical-analysis-of-the-ddos-attacks-against-ukrainian-websites/
        )
      </a>
    </h1>
    <div
      className="article-data-block"
    >
      <p
        className="article-data"
      >
        by: 
        idletom
        ,
      </p>
      <p
        className="article-data"
      >
        posted: 
        20.02.2022, 18:33
        ,
      </p>
      <p
        className="article-data"
      >
        comments: 
        0
      </p>
    </div>
    <hr />
    <div
      className="comment-block"
    >
      
    </div>
  </div>,
]
`;

exports[`Article screen renders correctly if error occured 1`] = `
Array [
  <div>
    <div
      className="header-controls"
    >
      <a
        className="header-controls-button header-controls-refresh-button "
        onClick={[Function]}
      />
      <a
        className="header-controls-button header-controls-return-button"
        href="/"
        onClick={[Function]}
      />
    </div>
    <p
      className="header"
    >
      Hacker News
    </p>
  </div>,
  <div
    className="container"
  >
    <div
      className="load"
    >
      <h1>
        Article does not exist
      </h1>
      <a
        href="/"
        onClick={[Function]}
      >
        Return to main page
      </a>
    </div>
  </div>,
]
`;
